New example: tracer in static mixer (#1362)
chaos-polymtl/lethe: doc/source/examples/multiphysics/tracer-in-static-mixer/tracer-in-static-mixer.rst

======================================
Tracer in Static Mixer
======================================

In this example, we extend the :doc:`../../sharp-immersed-boundary/3d-rbf-static-mixer/3d-rbf-static-mixer` example by using a passive tracer to study mixing effects more formally.

> We inject two phases through a static mixer to verify the mixing effects of its geometry.

Reviewer: ```suggestion
We inject two phases through a static mixer to verify how geometry affects their mixing.
```

+-----------------------------------------------------------------------------------------------------------------------------+
|  .. figure:: images/static_mixer_stl_casing_arrows.png                                                                      |
|     :align: center                                                                                                          |
|     :width: 800                                                                                                             |
|     :name: Surface grid representation of a helix static mixer with its casing.                                             |
|                                                                                                                             |
+-----------------------------------------------------------------------------------------------------------------------------+

----------------------------------
Features
----------------------------------

> - Solver: ``lethe-fluid-sharp``
> - Transient problem
> - Displays the use of the tracer physics
> - Usage of the ``average_velocity_profile`` initial condition to accelerate transient auxiliary physics resolution

Reviewer: ```suggestion
- Solver: ``lethe-fluid-sharp``
- Transient problem
- Tracer physics
- Accelerated transient auxiliary physics resolution with ``average_velocity_profile`` initial condition. 
```

---------------------------
Files Used in this Example
---------------------------

> All files mentioned below are located in the example's folder (``examples/multiphysics/tracer-in-static-mixer``).
> 
> * Parameter file for solving the flow: ``flow_in_long_mixer.prm``;
> * Parameter file for solving the tracer transport: ``flow_in_long_mixer_only_tracer.prm``;
> * Composite geometry file: ``mixer_long.composite``;
> * RBF geometry file: ``RBF_helix.output``. The extension is ``.output`` because it was named from a `bitpit <https://github.com/optimad/bitpit>`_ perspective. RBF preparation was done in the previous :doc:`../../sharp-immersed-boundary/3d-rbf-static-mixer/3d-rbf-static-mixer` example, based on this surface grid from Thingiverse [#thingiverse]_ under CC BY 4.0;
> * Python post-processing script: ``postprocess_tracer.py``.

Reviewer: ```suggestion
All mentioned files are found in the example's folder (``examples/multiphysics/tracer-in-static-mixer``).

* Parameter file for solving the flow: ``flow_in_long_mixer.prm``;
* Parameter file for solving the tracer transport: ``flow_in_long_mixer_only_tracer.prm``;
* Composite geometry file: ``mixer_long.composite``;
* RBF geometry file: ``RBF_helix.output``. The extension is ``.output`` because it was named from a `bitpit <https://github.com/optimad/bitpit>`_ perspective. RBF preparation was done in the previous :doc:`../../sharp-immersed-boundary/3d-rbf-static-mixer/3d-rbf-static-mixer` example, based on this surface grid from Thingiverse [#thingiverse]_ under CC BY 4.0;
* Python post-processing script: ``postprocess_tracer.py``.
```

#. Solving a pseudo-steady problem using a ``bdf1`` integration scheme with a small ``type step``, until a steady solution is reached;
#. Solving a transient tracer simulation with a large ``time step`` while reusing the computed velocity field from the previous simulation, which dramatically reduces computing time.

--------------
Parameter File
--------------

Simulation 1: Flow resolution
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Simulation control
******************

.. code-block:: text

    subsection simulation control
      set method    = bdf1
      set time end  = 40e-4
      set time step = 1e-4

      set adapt = false

      set output path      = ./output/
      set output name      = output
      set output control   = iteration
      set output frequency = 5
    end

#. We use a ``bdf1`` time integration scheme with a short ``time step = 1e-4``. After ``time end = 40e-4``, we consider that the velocity field has reached a steady-state.
#. We do not allow time step adaptation, since it is a parameter that we control ourselves between simulation: ``adapt = false``.

Restart
******************

> We use the checkpoint/restart mechanism of Lethe to achieve these simulations.

Reviewer: ```suggestion
We use Lethe's checkpoint/restart mechanism to feed the steady-state flow information to the second step of the simulation.
```

.. code-block:: text

    subsection restart
      set checkpoint = true
      set frequency  = 5
      set filename   = restart
      set restart    = false
    end

> #. We ``checkpoint`` the simulation at each ``5`` time steps, both for safety (each time step takes a long time to complete) and for reuse in the second simulation.
> #. We do not ``restart`` the simulation, unless it is stopped before it reaches ``time end``.

Reviewer: ```suggestion
#. We ``checkpoint`` the simulation at every ``5`` time steps, both for safety (each time step takes a long time to complete) and for reuse in the second simulation.
#. We do not ``restart`` the simulation unless stopped before it reaches ``time end``.
```

Multiphysics
******************

> Both ``fluid dynamics`` and ``tracer`` are enabed for the first simulation, although nothing happens tracer-wise.

Reviewer: ```suggestion
Both ``fluid dynamics`` and ``tracer`` are enabled for the first simulation. However, the tracer injection only begins at the second step.
```

Reviewer: I would mention why you need the tracer here if you don't use it.

.. code-block:: text

    subsection multiphysics
      set fluid dynamics = true
      set tracer         = true
    end

Physical Properties
*******************

The physical properties here are defined using centimeter as the length-unit and seconds as the time-unit. We consider that we have a passive tracer in water.

.. code-block:: text

    subsection physical properties
      subsection fluid 0
        set kinematic viscosity      = 0.01
        set tracer diffusivity model = immersed solid tanh
        subsection immersed solid tanh
          set tracer diffusivity inside  = 1e-10
          set tracer diffusivity outside = 1e-5
          set thickness                  = 5e-1
        end
      end
    end

> #. The ``tracer diffusivity model`` is ``immersed solid tanh``. This model is used for tracer simulations involving immersed solids when using ``lethe-fluid-sharp``.
> #. The ``tracer diffusivity outside`` is ``1e-5``, as this is a typical value for a passive tracer in a liquid.
> #. The ``tracer diffusivity inside`` is set to ``1e-10``. We desire no diffusivity inside the solid, and this value is low enough to achieve this effect while providing numerical stability.
> #. The ``thickness`` is ``5e-1``. At the scale of the problem, this provides a smooth transition without generating oscillations between liquid and solid phases.

Reviewer: ```suggestion
#. The ``tracer diffusivity model`` is ``immersed solid tanh``. This model is used in ``lethe-fluid-sharp`` for tracer flow percolating immersed solids.
#. The ``tracer diffusivity outside`` is ``1e-5``, as this is a typical value for a passive tracer in a liquid.
#. The ``tracer diffusivity inside`` is set to ``1e-10``. The low value prevents diffusivity inside the solid while providing numerical stability.
#. The ``thickness`` is ``5e-1``. At the scale of the problem, this provides a smooth transition without generating oscillations between liquid and solid phases.
```

Reviewer: Missing units.

Reviewer: Is this based on previous tests? It would be nice to mention why these values.

Author: It wasn't based on previous tests, more on value we thought were appropriate.

        set type = outlet
      end
    end

#. The boundary conditions are ``time dependent`` because of the inlet. We put conditions on :math:`y<0` and :math:`11 < t < 61` to inject a pulse, only on the lower half of the inlet.
#. All other boundary conditions are ``outlet``. This condition is natural for the outlet of the problem. For lateral walls, this condition is also appropriate since no flow is occurring through these; it is then equivalent to a impermeable wall.

Post-processing
*******************

.. code-block:: text

    subsection post-processing
      set verbosity = verbose

      set calculate average velocities = true
      set initial time                 = 30e-4

      # Tracer post-processing
      set calculate tracer flow rate  = true
      set tracer flow rate name       = tracer_flow_rate
    end

#. ``calculate average velocities`` is enabled, and its ``initial time = 30e-4``. This means that the last :math:`25\%` of the first simulation will be used to calculate the average velocity profile.
#. ``calculate tracer flow rate`` is enabled to provide data for understanding mixing effects.

Simulation 2: Tracer transport
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> In the second simulation, we reuse the velocity profile from the first simulation and simply transport the passive tracer. This allows to increase the time step and duration of the simulation while keeping realistic velocity and pressure solutions.

Reviewer: ```suggestion
In the second simulation, we reuse the velocity profile from the first simulation and simply transport the passive tracer. This allows us to increase the time step and duration of the simulation while keeping realistic velocity and pressure solutions.
```

Simulation control
******************

.. code-block:: text

    subsection simulation control
      set method    = bdf1
      set time end  = 500
      set time step = 10
      ...
    end

> #. We use a longer ``time step`` of ``10``.
> #. We let the simulation run until ``time end = 500`` to allow the tracer to pass through.

Reviewer: ```suggestion
#. We use a longer ``time step`` of ``10`` seconds.
#. We simulate until ``time end = 500`` seconds to allow the tracer to flow through the entire length of the domain.
```
